dlt/destinations/impl/synapse/synapse.py

```diff
@@ -0,0 +1,175 @@
+from typing import ClassVar, Sequence, List, Dict, Any, Optional, cast
+from copy import deepcopy
+from textwrap import dedent
+
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.common.destination.reference import SupportsStagingDestination, NewLoadJob
+
+from dlt.common.schema import TTableSchema, TColumnSchema, Schema, TColumnHint
+from dlt.common.schema.typing import TTableSchemaColumns, TTableIndexType
+
+from dlt.destinations.sql_jobs import SqlStagingCopyJob, SqlJobParams
+from dlt.destinations.sql_client import SqlClientBase
+from dlt.destinations.insert_job_client import InsertValuesJobClient
+from dlt.destinations.job_client_impl import SqlJobClientBase
+
+from dlt.destinations.impl.mssql.mssql import (
+    MsSqlTypeMapper,
+    MsSqlClient,
+    VARCHAR_MAX_N,
+    VARBINARY_MAX_N,
+)
+
+from dlt.destinations.impl.synapse import capabilities
+from dlt.destinations.impl.synapse.sql_client import SynapseSqlClient
+from dlt.destinations.impl.synapse.configuration import SynapseClientConfiguration
+
+
+HINT_TO_SYNAPSE_ATTR: Dict[TColumnHint, str] = {
+    "primary_key": "PRIMARY KEY NONCLUSTERED NOT ENFORCED",
+    "unique": "UNIQUE NOT ENFORCED",
+}
+TABLE_INDEX_TYPE_TO_SYNAPSE_ATTR: Dict[TTableIndexType, str] = {
+    "heap": "HEAP",
+    "clustered_columnstore_index": "CLUSTERED COLUMNSTORE INDEX",
+}
+
+
+class SynapseClient(MsSqlClient):
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(self, schema: Schema, config: SynapseClientConfiguration) -> None:
+        super().__init__(schema, config)
+        self.config: SynapseClientConfiguration = config
+        self.sql_client = SynapseSqlClient(
+            config.normalize_dataset_name(schema), config.credentials
+        )
+
+        self.active_hints = deepcopy(HINT_TO_SYNAPSE_ATTR)
+        if not self.config.create_indexes:
+            self.active_hints.pop("primary_key", None)
+            self.active_hints.pop("unique", None)
+
+    def _get_table_update_sql(
+        self, table_name: str, new_columns: Sequence[TColumnSchema], generate_alter: bool
+    ) -> List[str]:
+        table = self.get_load_table(table_name)
+        if table is None:
+            table_index_type = self.config.default_table_index_type
+        else:
+            table_index_type = table.get("table_index_type")
+            if table_index_type == "clustered_columnstore_index":
+                new_columns = self._get_columstore_valid_columns(new_columns)
```

Comment:
I think instead of doing this here, you should probably implement _get_column_def_sql (extend it to also accept index type) and react to it in the type_mapper. see how we do it with athena tables.

Reply:
Now that `table_index_type` (generic for all destinations) has been "downgraded" to `x-table-index-type` (Synapse-specific), would you still recommend this?

I ask because taking this approach would require fetching `x-table-index-type` from the table schema in `SqlJobClientBase._get_table_update_sql`, and then pass this value to `_get_column_def_sql(...)`, thus having to add `table_index_type` as an argument to all implementations for that function (similar to how `table_format` is handled). Destination-specific logic would end up in generic and other-destination-specific code (where it is unused).

Reply:
alright, let's keep it like this!